### PR TITLE
Implement basic screen rotation as requested in #297 

### PR DESCRIPTION
**Pull request by hrydgard**
Does not work correctly in non-buffered rendering yet. In fact, doing this is going to be quite tricky for non-buffered as we will need to rotate scissors, projection matrices etc, not sure it's worth it.

Also, it seems obvious that this should also rotate the controls somehow. Is it enough to rotate the D-pad and analog stick, or should we also rotate the four action buttons?

What to call the option is also an issue. Same as the Android screen rotation?

### CODE REVIEW

**Comment by sum2012** [comment]
We can disable the option of non-buffered rendering

**Comment by sum2012** [comment]
Well done

original
https://cloud.githubusercontent.com/assets/2177532/7217452/8d614cbc-e65e-11e4-8818-89e78ff239f5.jpg

changed:

File: GPU/GLES/Framebuffer.cpp [hunk 1]
```diff
@@ -533,15 +487,36 @@ void FramebufferManager::DrawPlainColor(u32 color) {
...
+	if (uvRotation != ROTATION_LOCKED_HORIZONTAL) {
+		float temp[8];
+		int rotation = 0;
+		switch (uvRotation) {
+		case ROTATION_LOCKED_HORIZONTAL180: rotation = 4; break;
+		case ROTATION_LOCKED_VERTICAL: rotation = 2; break;
+		case ROTATION_LOCKED_VERTICAL180: rotation = 6; break;
+		}
+		for (int i = 0; i < 8; i++) {
+			temp[i] = texCoords[(i + rotation) & 7];
```
Comment (unknownbrackets) [inline GPU/GLES/Framebuffer.cpp]: Nice.

-[Unknown]

**Comment by daniel229** [comment]
Great feature,it'd better to rotate the controls also the four action buttons and the layouts at the same time when rotate screen.

**Comment by unknownbrackets** [comment]
I guess ideally, it would also rotate the window size too?  Or maybe that's not really ideal.  Not sure, but if I'm playing windowed I guess I want 1x display to be 1x.  That being said, this is still better than not having the feature, and more users will (probably?) use full screen.

For the name... as is, maybe "Rendering rotation" or "Display rotation"?  Or "Rotate displayed screen"?  Probably bikeshedding, though, as long as it has "rotat*" and is in graphics it probably won't take too long for most users to figure out.

-[Unknown]

File: GPU/GLES/Framebuffer.cpp (hunk 1 above)
Comment (hrydgard) [inline GPU/GLES/Framebuffer.cpp]: Hehe, it is a little obscure but didn't feel like writing out all the cases :)

**Comment by hrydgard** [comment]
Not sure if it should change the window size directly, but I guess if you choose 1x or 2x from the menu when in rotated mode, it should use the vertical aspect ratio when autosizing the window.

I will add control rotation too shortly.

**Comment by i30817** [comment]
I assume there is no way to automatically recognize which games expect to be played rotated right?

**Comment by hrydgard** [comment]
@i30817 , no there isn't. Could always have a database of game IDs or something though.

**Comment by lightfirst** [comment]
enalbe postprocessing shader will broken screen rotation

**Comment by hrydgard** [comment]
Oh, thanks for reporting @lightfirst , I will look into it.

**Comment by hrydgard** [comment]
Actually, I just tried it and it seems to work just fine? What's the problem?

postwipeout

**Comment by lightfirst** [comment]
disable postprocessing shader
a

enalbe postprocessing shader(FXAA)
b

**Comment by lightfirst** [comment]
BTW, not support Linux, MAC?

**Comment by daniel229** [comment]
Postprocessing shader looks fine here.

**Comment by hrydgard** [comment]
@lightfirst  not initially, although it will partially work (not correct window resizing though)

Yeah, I can't get postprocessing to misbehave.. Can you post your ppsspp.ini, maybe you have some weird setting @lightfirst ?

**Comment by lightfirst** [comment]
my ppsspp.ini

[General]
FirstRun = False
RunCount = 2
Enable Logging = True
AutoRun = True
Browse = False
IgnoreBadMemAccess = True
CurrentDirectory = F:/umdimages/
ShowDebuggerOnLoad = False
CheckForNewVersion = True
Language = en_US
ForceLagSync = False
NumWorkerThreads = 3
EnableAutoLoad = False
EnableCheats = False
CwCheatRefreshRate = 77
ScreenshotsAsPNG = False
StateSlot = 0
RewindFlipFrequency = 0
GridView1 = True
GridView2 = True
GridView3 = False
ReportingHost = default
AutoSaveSymbolMap = False
CacheFullIsoInRam = False
InternalScreenRotation = 2
TopMost = False
WindowX = 165
WindowY = 63
WindowWidth = 976
WindowHeight = 603
PauseOnLostFocus = False
PauseWhenMinimized = False
DumpDecryptedEboots = False
[CPU]
Jit = True
SeparateCPUThread = False
SeparateIOThread = True
IOTimingMethod = 0
FastMemoryAccess = True
FuncReplacements = True
CPUSpeed = 0
SetRoundingMode = True
ForceFlushToZero = False
[Graphics]
EnableCardboard = False
CardboardScreenSize = 50
CardboardXShift = 0
CardboardYShift = 0
ShowFPSCounter = 0
GPUBackend = 0
RenderingMode = 1
SoftwareRendering = False
HardwareTransform = True
SoftwareSkinning = True
TextureFiltering = 1
BufferFiltering = 1
InternalResolution = 0
AndroidHwScale = 1
FrameSkip = 0
AutoFrameSkip = False
FrameRate = 0
FrameSkipUnthrottle = False
ForceMaxEmulatedFPS = 60
AnisotropyLevel = 4
VertexCache = True
TextureBackoffCache = False
TextureSecondaryCache = False
FullScreen = False
PartialStretch = False
StretchToDisplay = False
SmallDisplay = False
ImmersiveMode = False
TrueColor = True
MipMap = True
TexScalingLevel = 1
TexScalingType = 0
TexDeposterize = False
VSyncInterval = False
DisableStencilTest = False
AlwaysDepthWrite = False
DepthRangeHack = False
BloomHack = 0
TimerHack = False
AlphaMaskHack = False
SplineBezierQuality = 2
PostShader = Off
MemBlockTransferGPU = True
DisableSlowFramebufEffects = False
FragmentTestCache = True
[Sound]
Enable = True
AudioBackend = 0
AudioLatency = 1
SoundSpeedHack = False
AudioResampler = True
[Control]
HapticFeedback = True
ShowTouchCross = True
ShowTouchCircle = True
ShowTouchSquare = True
ShowTouchTriangle = True
ShowTouchStart = True
ShowTouchSelect = True
ShowTouchLTrigger = True
ShowTouchRTrigger = True
ShowAnalogStick = True
ShowTouchDpad = True
ShowTouchUnthrottle = True
ShowTouchPause = True
IgnoreWindowsKey = False
ShowTouchControls = False
DisableDpadDiagonals = False
GamepadOnlyFocused = False
TouchButtonStyle = 1
TouchButtonOpacity = 65
AutoCenterTouchAnalog = False
ActionButtonSpacing2 = 1.000000
ActionButtonCenterX = 0.875000
ActionButtonCenterY = 0.779412
ActionButtonScale = 1.150000
DPadX = 0.148958
DPadY = 0.470588
DPadScale = 1.150000
DPadSpacing = 1.000000
StartKeyX = 0.619792
StartKeyY = 0.873162
StartKeyScale = 1.150000
SelectKeyX = 0.500000
SelectKeyY = 0.873162
SelectKeyScale = 1.150000
UnthrottleKeyX = 0.380208
UnthrottleKeyY = 0.873162
UnthrottleKeyScale = 1.150000
LKeyX = 0.083333
LKeyY = 0.084559
LKeyScale = 1.150000
RKeyX = 0.928125
RKeyY = 0.084559
RKeyScale = 1.150000
AnalogStickX = 0.148958
AnalogStickY = 0.830882
AnalogStickScale = 1.150000
DInputAnalogDeadzone = 0.100000
DInputAnalogInverseMode = 0
DInputAnalogInverseDeadzone = 0.000000
DInputAnalogSensitivity = 1.000000
XInputAnalogDeadzone = 0.240000
XInputAnalogInverseMode = 0
XInputAnalogInverseDeadzone = 0.000000
XInputAnalogSensitivity = 1.000000
AnalogLimiterDeadzone = 0.600000
[Network]
EnableWlan = False
EnableAdhocServer = False
[SystemParam]
PSPModel = 1
PSPFirmwareVersion = 150
NickName = PPSSPP
proAdhocServer = coldbird.net
MacAddress = 0c:5b:98:16:ea:b6
Language = 1
TimeFormat = 1
DateFormat = 1
TimeZone = 0
DayLightSavings = False
ButtonPreference = 1
LockParentalLevel = 0
WlanAdhocChannel = 0
BypassOSKWithKeyboard = False
WlanPowerSave = False
EncryptSave = True
[Debugger]
DisasmWindowX = 173
DisasmWindowY = 51
DisasmWindowW = 781
DisasmWindowH = 713
GEWindowX = 173
GEWindowY = 75
GEWindowW = 776
GEWindowH = 699
ConsoleWindowX = -1
ConsoleWindowY = -1
FontWidth = 8
FontHeight = 12
DisplayStatusBar = True
ShowBottomTabTitles = True
ShowDeveloperMenu = False
SkipDeadbeefFilling = False
FuncHashMap = False
[SpeedHacks]
PrescaleUV = False
DisableAlphaTest = False
[JIT]
[Upgrade]
UpgradeMessage = 
UpgradeVersion = 
DismissedVersion = 
[Recent]
MaxRecent = 30
FileName0 = F:/umdimages/PSP GVG next plus.ISO
[PinnedPaths]

**Comment by daniel229** [comment]
That setting is fine.Maybe the .exe build problem.
@lightfirst Try this one to see if good for you.

Rename jpg to 7z
ppsspp-v1 0 1-485-g1b68987-windows

**Comment by lightfirst** [comment]
@daniel229 it's good using you .exe build. thinks

**Pull request merged**